Inference Cost proposal (#1646)
llm-d/llm-d: docs/proposals/inference-costs.md

- **Multi-tenant cost allocation**: Cannot fairly distribute infrastructure costs across multiple teams and workloads sharing the same model servers
- **GPU Efficiency Opacity**: Limited ability to right size underlying GPU infrastructure to actual observed workload resource requirements 

### Goals

1. **Infrastructure-based cost calculation**: Calculate per-token costs from actual infrastructure use (ex: GPU, CPU, memory infrastructure usage) not pre-configured pricing tables
2. **Multi-dimensional attribution**: Track costs by team, workload, model, model variant, label and namespace
3. **Differentiated token pricing**: Provide separate costs for input (prompt) and output (generation) tokens based on actual compute time
4. **Seamless integration**: Integrate with existing llm-d metrics ecosystem (DCGM, vLLM, EPP, GPU metrics) without duplication
5. **Disaggregation support**: Track costs for disaggregated serving (prefill/decode) deployments
6. **Production-ready**: Provide reliable, scalable cost tracking suitable for production deployments
7. **Optimization insights**: Enable measurement of cost savings from KV cache, prefix caching, and other optimizations

> 8. **LLM/GPU workload idle insights**: Improve on OpenCosts current GPU capabilities to better identify GPU idle

Reviewer: ```suggestion
8. **LLM/GPU workload idle insights**: Improve on OpenCost's current GPU capabilities to better identify GPU idle
```

Note: Although not a direct goal of this project, collection of SaaS Model Provider usage and billing can be obtained via OpenCost via its plugin mechanism.  This makes comparing self hosted costs to SaaS model costs much easier.  Currently there is a [plugin for OpenAI](https://github.com/opencost/opencost-plugins/tree/main/pkg/plugins/openai).

### Non-Goals

1. **Not replacing existing metrics**: This proposal does not replace llm-d's existing performance and operational metrics
2. **Not a billing system**: This is not a billing/invoicing system, but provides cost data for such systems
3. **Not for training costs**: Focus is exclusively on inference costs, not model training nor fine-tuning.  Changes made to OpenCost, however, will be made with an eye towards enabling future work such as training cost support. 
4. **Not real-time billing**: Cost calculations are based on recent metrics (5-minute windows), not per-request billing
5. **Not pricing**: This proposal provides costs not prices.  If static or dynamic pricing is desired it can be generated using inference costs as a basis.

## Cost Metrics Primer

> For background on infrastructure cost concepts (alloccation-based vs usage-only costs, idle, wasted, and shared costs), see the [Cost Metrics Primer](./cost-metrics-primer.md).  

Reviewer: ```suggestion
For background on infrastructure cost concepts (allocation-based vs usage-only costs, idle, and shared costs), see the [Cost Metrics Primer](./cost-metrics-primer.md).  
```

_This is important for understanding the following sections._

### User Stories

#### Story 1: Platform Team Cost Optimization

As an inference platform team managing multiple AI model and/or llm-d deployments, I want to track infrastructure costs per model and variant so I can identify which configurations are most cost-effective and make data-driven decisions about resource allocation.

**Acceptance Criteria**:

> - View cost per million tokens for each model

Reviewer 2: Are these input tokens, output tokens, cached tokens? All three?

```

### Core Approach

In addition to collecting and calculating general infrastructure costs as is already done by OpenCost today, we will add inference specific cost metrics.

1. **Collect metrics from existing sources**:
   - Token metrics from vLLM (ex: `vllm:prompt_tokens_total`, `vllm:generation_tokens_total`)
   - GPU costs from OpenCost's existing allocation system
   - GPU requests from DRA and device plugins
   - GPU utilization 
   - Processing time metrics from vLLM for differentiated pricing
   - KV cache hits and usage metrics
   - Namespace and model labels for attribution

2. **Calculate infrastructure-based costs**:
   - Determine total cost per container based on allocation and node costs
   - Calculate total tokens processed in time window
   - Compute cost per token from infrastructure costs divided by token throughput as well as based on usage
   - Allocate costs between input and output tokens based on actual processing time
   - Calculate cache saving due to optimizations such as KV cache hits, prefill/decode separation, and smart routing
   - Identify idle costs - a key metric needed to enable optimization of infrastructure 

>    - Provide both usage and reservation cost basis metrics

Reviewer: ```suggestion
   - Provide both usage and allocation cost basis metrics
```

4. **Enable multi-dimensional queries**:
   - Aggregate by model, namespace, team/tenant, workload
   - Filter by time windows
   - Compare costs across different configurations - i.e. model variants
   - Monitor model and model variant idle costs

## Design Details

The proposal is to:
1. Define and standardize inference cost metrics
2. Provide a default implementation based on OpenCost that generates and provides access to the metrics

The goal is to provide a fully functional cost solution for llm-d using OpenCost, but allowing for it to be replaced as long as the replacement for OpenCost is able to generate the same metrics and provides the same APIs.

### Implementation Status

The proposal is based on a **working proof of concept**, which demonstrates:

- Basic cost metrics (total cost, cost per million tokens)
- Differentiated input/output costs with compute-time allocation - simple, gpu only, no kv cache calculations

The [POC implementation](https://github.com/simanadler/opencost/tree/initial-inference) validates the overall approach and feasibility of using OpenCost, and does not aim to provide fully accurate costs.

## High Level Design Proposal

### LLM costs

> _For background on infrastructure cost concepts (reservation-based vs usage-only costs, idle, wasted, and shared costs), see the [Cost Metrics Primer](cost-metrics-primer.md)._

Reviewer: ```suggestion
_For background on infrastructure cost concepts (allocation-based vs usage-only costs, idle, wasted, and shared costs), see the [Cost Metrics Primer](cost-metrics-primer.md)._
```

- **Wasted** — GPU memory holding model weights while no requests are arriving (model loaded but idle between requests)
- **Idle** — GPUs with no model loaded at all or with a model but requesting more resources than required
- **Shared** — Typical kubernetes shared costs **plus kv cache** and other inference specific shared costs described [here](#beyond-just-gpu)

The gap between "used" and "wasted" can be far larger for LLMs than for CPU workloads. A CPU workload can be bin-packed with others to fill spare capacity. An LLM model loaded onto a GPU blocks that GPU entirely — its weights occupy VRAM whether or not a request is in flight, and loading/unloading is slow and expensive. A low-traffic model may spend 95% of its time in the "wasted" state.

The following are the proposed metrics to be added to support AI Inferencing.  

**Proposed LLM Cost Metrics**:

- `llm_total_cost`: Hourly cost per model
- `llm_cost_per_million_tokens`: Blended cost per 1M tokens - without kv cache savings
- `llm_input_cost_per_million_tokens`: Cost per 1M **input tokens** - without kv cache savings
- `llm_output_cost_per_million_tokens`: Cost per 1M **output** tokens

> - `llm_total_kv_cache_savings`: Hourly savings due to KV cache hits

Reviewer 2: It would be great if this can be expressed in the USD per 1M cached tokens to make this directly comparable with SaaS

- `llm_kv_cache_savings_per_million_tokens`: KV cache savings per million tokens

**Cost Metric Labels**: 

New labels:
* model_name="meta-llama/Llama-3.1-8B-Instruct" (example)
* model_version="v1.0" (example)
* model_variant="prefill" (example)

> * cost_basis="usage" or "reservation"

Reviewer: ```suggestion
* cost_basis="usage" or "allocation"
```

**Usage-based cost per model** counts only the GPU compute consumed during active inference — the tokens actually generated minus costs saved by kv cache hits. This answers: *"what did this model's actual work cost?"* The gap between this and the reservation-based figure is the cost of keeping the model warm and ready, which is a business decision about acceptable latency, not a technical inefficiency.

### Cost Per Million Tokens 

Both cost metrics produce a corresponding cost per million tokens figure, but they answer different questions.

**Usage-based Cost per Million Tokens** measures the intrinsic cost of inference — what each token costs in pure compute terms. This is stable regardless of how busy the model is, making it the right metric for comparing the efficiency of different models or hardware configurations against each other, or against external API pricing for the inference work itself.

> **Reservation-based Cost Per Million Tokens** measures the real business cost of each token when you factor in the full price of keeping the model available, including idle time. This varies with utilization — the same model serving fewer requests will show a higher reservation-based cost per token because the fixed hosting cost is spread across fewer tokens.

Reviewer: ```suggestion
**Allocation-based Cost Per Million Tokens** measures the real business cost of each token when you factor in the full price of keeping the model available, including idle time. This varies with utilization — the same model serving fewer requests will show a higher allocation-based cost per token because the fixed hosting cost is spread across fewer tokens.
```

The relationship between the two directly expresses utilization without needing a separate metric:

```

> utilization = usage-based cost per million tokens / reservation-based cost per million tokens

Reviewer: ```suggestion
utilization = usage-based cost per million tokens / allocation-based cost per million tokens
```

Example:
  Usage-based:       $1.00 per million tokens  (intrinsic inference cost)

>   Reservation-based: $4.00 per million tokens  (full hosting cost)

Reviewer: ```suggestion
  Allocation-based: $4.00 per million tokens  (full hosting cost)
```

  Utilization:       25%  (model is busy 1/4 of the time)
```

These metrics will be broken down further into the cost per million **input** and **output** tokens.  In order to calculate input token cost it is necessary to capture kv cache hit savings, which will also be provided as a metric.

**Cost Per Million Tokens - with KV Cache Hits**

To calculate the actual cost per million tokens its necessary to do the following calculation:

```actual_cost_per_million_tokens = llm_cost_per_million_tokens - llm_kv_cache_savings_per_million_tokens```

or if you want specifically for input tokens only:

> ```actual_cost_per_million_input_tokens = llm_cost_per_million_input_tokens - llm_kv_cache_savings_per_million_tokens```

Reviewer 2: I think this is the same calculation, but just to make sure. In SaaS: `actual_cost_per_million_input_tokens = (base_cost * (input_tokens_uncached  + input_tokens_cached  (1 - discount)))/1M`, where `discount` is a real number in `[0,1]`

|---|---|---|---|
| **Inference Scheduler / EPP** | `inferencepool` (gateway-api-inference-extension) | **1:1 with InferencePool*** | CPU-only; routes all traffic for the pool |
| **llm-d-infra / Gateway proxy** | `llm-d-infra` | **1:1 with InferencePool** | Sets up gateway; may include a proxy pod |
| **Async Processor** | `async-processor` | **1:1 with InferencePool** | Present only in async-processing deployments; points to single gateway URL |
| **Workload Variant Autoscaler (WVA)** | `workload-variant-autoscaler` | **Cluster-wide by default** | Watches all namespaces unless `namespaceScoped: true`; sole cross-pool component |
| **KV-cache PVC** | manual PVC / `llm-d-kv-cache-storage` | **Per model or shared across model variants** | Can be up to 18TB in tiered-cache deployments |

    Note: An InferencePool is an llm-d resource that groups a set of model server replicas (pods) behind a single logical endpoint. It is the unit of deployment for a model in llm-d: the Inference Scheduler (EPP), gateway proxy, and other per-model infrastructure components are each scoped 1:1 to an InferencePool.  Different models within an InferencePool are referred to as model variants.

>   **Recommendation**: Treat KV cache as a model reservation cost for the purposes of cost attribution. 

Reviewer: ```suggestion
  **Recommendation**: Treat KV cache as a model allocation cost for the purposes of cost attribution. 
```

    The reason is that KV cache capacity is sized at deployment time based on expected concurrency, not actual requests — the operator makes a deliberate choice to reserve X GB for KV cache when deploying the model. Charging it to usage would obscure that capacity decision and make usage-based costs misleadingly low for
    memory-intensive models.

    There are situations where portions of the KV cache are offloaded to disk when VRAM is exhausted.  This memory is consumed in proportion to actual request load — more concurrent requests means more KV cache pages in use. This is closer to a usage cost since it scales with actual inference activity - but we will defer addressing this initially.

    In deployments where llm-d's architecture separates prefill and decode with a shared KV cache, the models sharing the KV cache are variants of the same model - i.e. they have different hardware and/or LoRa configurations but use the same model and version.  Thus, from a cost perspective we view them as a single model.
  
  
### The useful combination

The two per-model cost figures and the two per-token figures together give a complete picture:

> | Reservation cost | Usage-based cost per million tokens | What it means |

Reviewer: ```suggestion
| Allocation cost | Usage-based cost per million tokens | What it means |
```

|------------------|--------------------------------------|---------------|
| High             | Low                                  | Costly to keep available, but efficient at inference — utilization is the problem, consider shared serving |
| High             | High                                 | Costly to keep available *and* expensive to run — evaluate whether the model is the right choice |
| Low              | Low                                  | Cheap to host and efficient — well-sized deployment |
| Low              | High                                 | Cheap to host but the inference itself is costly — look at model size, quantization, or hardware fit |

### Comparing against external API pricing

> When deciding whether to self-host or use an external API, the comparison must be made against **reservation-based cost per million tokens**, not usage-based. Usage-based cost per token reflects only active compute and will almost always make self-hosting look cheaper than it is. The reservation-based figure captures the full cost of availability — which is exactly what an external API provider is also charging you for.

Reviewer: ```suggestion
When deciding whether to self-host or use an external API, the comparison must be made against **allocation-based cost per million tokens**, not usage-based. Usage-based cost per token reflects only active compute and will almost always make self-hosting look cheaper than it is. The reservation-based figure captures the full cost of availability — which is exactly what an external API provider is also charging you for.
```

```
Self-hosted model:
  Usage-based cost per million tokens:       $1.00  (compute only)

>   Reservation-based cost per million tokens: $4.00  (full hosting cost)

Reviewer: ```suggestion
  Allocation-based cost per million tokens: $4.00  (full hosting cost)
```

External API price per million tokens:       $2.00

> Conclusion: external API is cheaper — self-hosting at 25% utilization costs $4.00/M tokens,

Reviewer: ```suggestion
Conclusion: External API is cheaper — self-hosting at 25% utilization costs $4.00/M tokens,
```
